engine/src/main/java/net/jqwik/engine/support/combinatorics/CombinedIterator.java

```diff
@@ -2,62 +2,79 @@
 
 import java.util.ArrayList;
 import java.util.*;
-
-import static java.util.Arrays.*;
+import java.util.stream.*;
 
 public class CombinedIterator<T> implements Iterator<List<T>> {
-
-	private final Iterator<T> first;
-	private final ArrayList<Iterable<T>> rest;
-	private Iterator<List<T>> next;
-
-	private T current = null;
-
-	// This must be tracked because there can be null values
-	private boolean currentIsSet = false;
+	private final List<Iterable<T>> iterables;
+	private final List<Iterator<T>> iterators;
+	private final List<T> elements;
+	private final boolean isEmpty;
+	private int position = -1;
 
 	public CombinedIterator(List<Iterable<T>> iterables) {
-		this.rest = new ArrayList<>(iterables);
-		this.first = this.rest.remove(0).iterator();
-		this.next = restIterator();
-	}
-
-	private Iterator<List<T>> restIterator() {
-		return this.rest.isEmpty()
-				   ? emptyListIterator()
-				   : new CombinedIterator<>(this.rest);
-	}
-
-	private Iterator<List<T>> emptyListIterator() {
-		return asList(Collections.<T>emptyList()).iterator();
+		this.iterables = iterables;
+		elements = new ArrayList<>(Collections.nCopies(iterables.size(), null));
+		iterators = iterables.stream().map(Iterable::iterator).collect(Collectors.toCollection(ArrayList::new));
+		isEmpty = !iterators.stream().allMatch(Iterator::hasNext);
```

Comment:
@jlink , I just noticed `allMatch` would return `true` for the empty stream, so the expression should be adjusted as `isEmpty = iterables.isEmpty() || !iterators.str...`